### PR TITLE
Autocut improvement

### PR DESCRIPTION
**Pull request by zeyueN**
**Context:**
Improves automatic cutting for qcut with a better way of deriving the imbalance parameter based on per-wire gate-depth of the circuit.

**Description of the Change:**
Changes to the signature and logic for `CutStrategy._infer_imbalance`. Previously it generates very large imbalance which will be ignored by kahypar internally. Now the imbalance are tighter which actually improves the performance.

Also added a test for auto cutting a MPS circuit, expecting optimal cut.

**Benefits:**
Better auto cut.

**Possible Drawbacks:**
There may exist circuits which the improved auto cut generates worse result. But so far I haven't seen one. Would be great if it can be tested on more types of circuits.

**Related GitHub Issues:**

### CODE REVIEW

**Comment by zsyr** [comment]
Hello, 

I am using Pennylane cut_circuit() library in Braket notebook instance. I ran into some issues of auto-cutting complicated circuits, e.g. VQE circuits. I want to try the improved auto-cutting feature that you developed in this page. How can I integrate the new feature into Braket notebook instance?

Thank you,
S.

**Comment by trbromley** [comment]
Hi @zsyr! To access this feature now, you can install the development version of PennyLane that follows this branch by running the following in the terminal:
```
pip install git+https://github.com/PennyLaneAI/pennylane.git#autocut_improvement
```
In the meantime, @anthayes92 and myself will prioritize reviewing this pull request, so that we can merge it and you can use the standard development version of PennyLane :+1:

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -208,6 +208,10 @@ of operators. [(#2622)](https://github.com/PennyLaneAI/pennylane/pull/2622)
 
 <h3>Improvements</h3>
 
+* The automatic circuit cutting is improved by making better partition imbalance derivations.
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:211]:
```suggestion
* Automatic circuit cutting is improved by making better partition imbalance derivations.
```

File: pennylane/transforms/qcut.py [hunk 2]
```diff
@@ -2434,8 +2436,10 @@ def _infer_imbalance(
             balancing_adjustment = 2 if avg_fragment_gates > 5 else 0
             free_gates = free_gates - (k - 1 + balancing_adjustment)
 
-        gate_imbalance = free_gates / avg_fragment_gates - 1
-        imbalance = max(gate_imbalance, 0.1 / avg_fragment_gates)  # numerical stability
+        depth_imbalance = max(wire_depths.values()) * num_wires / num_gates - 1
+        max_imbalance = free_gates / avg_fragment_gates - 1
+        # max_imbalance = max(max_imbalance, 0.1 / avg_fragment_gates)  # numerical stability
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:2441]:
```suggestion

```

File: tests/transforms/test_qcut.py [hunk 3]
```diff
@@ -4497,6 +4495,60 @@ def test_find_and_place_cuts(self, local_measurement, with_manual_cut, cut_strat
                 in expected_num_cut_edges
             )
 
+    @pytest.mark.parametrize("measure_all_wires", [False, True])
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:4498]: Should this go as an integration test in `TestAutoCutCircuit`?

File: tests/transforms/test_qcut.py [hunk 4]
```diff
@@ -4497,6 +4495,60 @@ def test_find_and_place_cuts(self, local_measurement, with_manual_cut, cut_strat
...
+
+        tape = tape0.expand()
+        graph = qcut.tape_to_graph(tape)
+        cut_graph = qcut.find_and_place_cuts(
+            graph=graph,
+            cut_strategy=cut_strategy,
+            replace_wire_cuts=True,
+        )
+        frags, _ = qcut.fragment_graph(cut_graph)
+        assert len(frags) == 7
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:4550]: Should we check that each fragment is of at most 2 qubits?

**Review by trbromley — CHANGES_REQUESTED** [review]
Thanks @zeyueN! Just left some minor suggestions but otherwise will be happy to approve!

File: tests/transforms/test_qcut.py [hunk 5]
```diff
@@ -4497,6 +4495,60 @@ def test_find_and_place_cuts(self, local_measurement, with_manual_cut, cut_strat
...
+            qml.RY(weights[0], wires=wires[0])
+            qml.RY(weights[1], wires=wires[1])
+
+        n_wires = 8
+        n_block_wires = 2
+        n_params_block = 2
+        n_blocks = qml.MPS.get_n_blocks(range(n_wires), n_block_wires)
+        template_weights = [[0.1, -0.3]] * n_blocks
+
+        cut_strategy = qml.transforms.qcut.CutStrategy(max_free_wires=2)
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:4533]: Curious what would happen without this constraint?

Comment (zeyueN) [reply tests/transforms/test_qcut.py:4533]: This is a required argument that tells the cutter how small the device is. It is only optional when the `devices` argument is available and passed.

**Review by trbromley — APPROVED** [review]
Thanks @zeyueN! Looks good, tested this and the automatic placement seems to work nicely!

**Pull request merged**